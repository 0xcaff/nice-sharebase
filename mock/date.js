// Sets any calls to date to return a constant time with a lot of zeros.
const constantDate = new Date(1500000000000);

global.Date.now = () => constantDate;