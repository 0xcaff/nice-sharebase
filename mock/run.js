require('babel-register');
require("regenerator-runtime/runtime");

const app = require('./server');
app.listen(30101, '0.0.0.0', function() { console.log('started mock server on port 30101') });
