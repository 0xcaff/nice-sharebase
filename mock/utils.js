import fs from 'fs';
import url, { URLSearchParams } from 'url';
import path from 'path';

// Returns the result of the first value in elems send through predicate which didn't throw an error.
export async function first(elems, predicate) {
	for (let elem of elems) {
		const { result, error } = await tryIt(_ => predicate(elem));
		if (!error) {
			return result;
		}
	}
}

// Wraps fs.readFile to use promises.
export const readFile = (...args) =>
  new Promise((resolve, reject) =>
    fs.readFile(...args, (err, data) => {
      if (err) { reject(err) } else { resolve(data) }
    })
  );

// Async returns an object with an error if one was caught. Otherwise returns the
// result.
export const tryIt = async (it) => {
  try {
    return {result: await it()};
  } catch(e) {
    return {error: e};
  }
};

// If a folder is requested with embeddings added, add we add them here.
export async function embedIfNeeded(folder, location) {
  const params = new URLSearchParams(location.search);
  const embed = params.get('embed');

  if (!embed) {
    return folder;
  }

  const embeds = embed.split(',');
  folder = JSON.parse(folder);

  const id = folder['FolderId'];
  if (embeds.includes('f')) {
    // try to embed folders
    const file = path.resolve(__dirname, `api/folders/${id}/folders.json`);

    const { result, error } = await tryIt(_ => readFile(file));
    if (!error && result) {
      const folders = JSON.parse(result);
      folder["Embedded"]["Folders"].push(...folders);
    }
  }

  // try to get folders
  if (embeds.includes('d')) {
    // try to embed documents
    const file = path.resolve(__dirname, `api/folders/${id}/documents.json`);

    const { result, error } = await tryIt(_ => readFile(file));
    if (!error && result) {
      const documents = JSON.parse(result);
      folder["Embedded"]["Documents"].push(...documents);
    }
  }

  return JSON.stringify(folder);
};

export async function handle(p) {
  const location = url.parse(p);
  p = location.pathname.slice(1);

	const paths = [`${p}.json`, `${p}/index.json`]
 		.map(raw => path.resolve(__dirname, raw));

  const result = await first(paths, path => readFile(path));
  if (result !== undefined) {
    const status = 200;
    const message = await embedIfNeeded(result, location);
    const headers = {'Content-Type': 'application/json'};

    return { status, message, headers };
  } else {
    const status = 404;
		const pre = `Path not found for ${p}. Tried:`;
		const tried = paths.map(path => '\t' + path).join('\n');
    const message = pre + '\n' + tried;

    return { status, message };
  }
}
