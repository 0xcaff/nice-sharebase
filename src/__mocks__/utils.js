const utils = require.requireActual('../utils');

utils.token = jest.fn(() => {
  const calls = utils.token.mock.calls.length;
  return `Session${calls}`;
});

module.exports = utils;
