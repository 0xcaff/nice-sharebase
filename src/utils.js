import crypto from 'crypto';

const get = (prop) => (obj) => obj[prop];

// Creates a new object copying all functions and replacing other values with
// getter functions.
export const getters = (kvs) =>
  Object.entries(kvs)
    .reduce(
      (acc, [key, value]) => {
        if (value instanceof Function) {
          // just copy it
          acc[key] = value;
        } else {
          acc[key] = get(value);
        }
        return acc;
      },
      {},
    );

// Returns a base64 encoded string of the string passed in.
export const base64Encode = (toEncode) => Buffer.from(toEncode).toString('base64');
export const base64Decode = (encoded) => Buffer.from(encoded, 'base64').toString();

// A function which returns its first argument.
export const nop = (arg) => arg;

// Copys the properties and their values of the input object to a new output
// object and returns it.
export const filter = (obj) =>
  Object.entries(obj)
    .reduce(
      (acc, [k, v]) => { if (v) { acc[k] = v } return acc; },
      {},
    );

// Creates a cryptographically random sequence of characters.
export const token = async () => base64Encode(await rand(256));

// Wraps the crypto.randomBytes API to use promises.
const rand = (size) =>
  new Promise((resolve, reject) =>
    crypto.randomBytes(size, (err, buf) => {
      if (err) { reject(err) }
      resolve(buf);
    })
  );

// A functional way to throw errors. This is useful for throwing errors when
// destructuring an object when a value isn't provided.
export const toss = (err) => { throw err };
export const required = (what, foor) =>
  toss(new TypeError(`${what} is a required parameter to ${foor}`));

export const MS = 1;
export const SECONDS = 1000 * MS;
export const MINUTES = 60 * SECONDS;
export const HOURS = 60 * MINUTES;
