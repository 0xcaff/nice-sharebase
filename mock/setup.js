// Sets up nock to respond to requests to the mock endpoint.
import nock from 'nock';

import { handle, authenticate } from './utils';

export const DEFAULT_ENDPOINT = 'http://mock.com/';

// TODO: move this log wrangling
export const logs = [];

global.beforeEach && beforeEach(() => logs.length = 0);

const promisify = (inner) =>
  function promiseWrapper(url, body, cb) {
    inner(this.req)
      .then((...args) => cb(null, ...args))
      .catch(err => cb(err));
  };

const translate = (inner) =>
  promisify(async req => {
    log(req.path);
    const { status, message, headers } = await inner(req);
    return [ status, message, headers ];
  });

const log = (path) => logs.push(path);

// Setups nock to intercept and handle HTTP requests.
export const setup = ({
  // The base url at which we will capture all traffic to.
  endpoint = DEFAULT_ENDPOINT,

  // Whether or not the authentication is required before accessing data. This
  // is used when testing.
  authEnabled = true,

  // The set of email:password allowed to access to get access tokens to this
  // site.
  users = new Set(['sam.babic@onbase.com:password']),

  // The set of allowed access tokens.
  allowed = new Set(),
} = {}) => {
  const app = nock(endpoint);
  app
    .get('/api/authenticate')
    .times(Infinity)
    .reply(translate(async req => authenticate(req.headers, allowed, users)));

  app
    .get(() => true)
    .times(Infinity)
    .reply(translate(req => handle(req.path, req.headers, allowed, authEnabled)));

  return app;
};