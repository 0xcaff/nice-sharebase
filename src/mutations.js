import url from 'url';
import 'isomorphic-fetch';

import { base64Encode, token } from './utils';
import { throwOnFail } from './errors';

// TODO: test me endpoint
// TODO: wire up proper authentication for everything

export const auth = async (email, password, { base, transform, sessionStore }) => {
  const inner = base64Encode(`${email}:${password}`);
  const headers = new Headers({'Authorization': `Basic ${inner}`});

  // According to the official API documentation, this should be a GET
  // request.

  // TODO: It's time to abstract the network layer away. Also log requests in
  // the app network layer instead of the mock network layer.
  const endpoint = url.resolve(base, 'authenticate');
  const request = new Request(endpoint, { headers: headers });
  const resp = await throwOnFail(await fetch(transform(request)));
  const res = await resp.json();

  // create a new session
  const delegate = await token();

  sessionStore.set(delegate, {
    expires: +Date.parse(res['ExpirationDate']),
    authed: res,
    email,
    password,
  });
  
  return { ...res, Token: delegate };
};
