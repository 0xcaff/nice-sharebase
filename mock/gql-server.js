import gql from 'express-graphql';
import express from 'express';
import url from 'url';
import { schema, createContext } from '../index.js';

import { ENDPOINT, logs } from './setup';

const app = express();

app.use('/graphql', gql(
  async (req, res, params) => {
    const base = url.resolve(ENDPOINT, 'api/');
    const context = createContext({ base });

    return {
      schema,
      context,
      extensions: (document, variables, operationName, result) => {
        const l = logs.slice();
        logs.length = 0;
        return l;
      },
      graphiql: true,
      pretty: true,
    };
  }
));

app.listen(3010, '0.0.0.0', _ => console.log('listening on 0.0.0.0:3010'));
