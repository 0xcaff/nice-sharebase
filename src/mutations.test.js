jest.mock('./utils');
import '../mock/date';

import url from 'url';
import { DEFAULT_ENDPOINT, setup } from '../mock/setup';
import { auth } from './mutations';
import { nop } from './utils';

const base = url.resolve(DEFAULT_ENDPOINT, 'api/');
setup();

it('should authenticate with the backend', async () => {
  const sessionStore = new Map();
  const resp = await auth('sam.babic@onbase.com', 'password',
    { base, transform: nop, sessionStore });

  expect(resp).toMatchSnapshot();
  expect(sessionStore).toMatchSnapshot();
});