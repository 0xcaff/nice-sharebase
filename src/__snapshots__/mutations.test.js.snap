// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should authenticate with the backend 1`] = `
Object {
  "ExpirationDate": "2017-07-14T02:50:00.000Z",
  "Token": "Session2",
  "UserId": 0,
  "UserName": "sam.babic@onbase.com",
}
`;

exports[`should authenticate with the backend 2`] = `
Map {
  "Session2" => Object {
    "authed": Object {
      "ExpirationDate": "2017-07-14T02:50:00.000Z",
      "Token": "Session1",
      "UserId": 0,
      "UserName": "sam.babic@onbase.com",
    },
    "email": "sam.babic@onbase.com",
    "expires": 1500000600000,
    "password": "password",
  },
}
`;
